Improve organization of BigQuery access cookbook (#478)
mozilla/firefox-data-docs: src/cookbooks/bigquery/access.md

which lists the tables and views in a BigQuery dataset:

```bash
bq ls moz-fx-data-derived-datasets:telemetry
```

And here's another which gets the count of entries in `telemetry.main` on `2019-08-22` in the nightly channel:

```bash
bq query --nouse_legacy_sql 'select count(*) from `moz-fx-data-derived-datasets.telemetry.main` where submission_date = "2019-08-22" and normalized_channel="nightly"'
```

Additional examples and documentation can be found [in the BigQuery command-line reference](https://cloud.google.com/bigquery/docs/bq-command-line-tool).

### API Access

> This method requires [BigQuery Access](#BigQuery-Access-Request) to be provisioned.

For advanced use cases involving programmatic access -- including automated workloads, ETL, [BigQuery Storage API](https://cloud.google.com/bigquery/docs/reference/storage/).

You can locate a list of supported BigQuery client libraries [here](https://cloud.google.com/bigquery/docs/reference/libraries).

Although usually you will want to use a client library, Google also provides a [detailed reference of their underlying REST API](https://cloud.google.com/bigquery/docs/reference/rest/).

> #### Service Accounts

Author: This should really be a seperate cookbook, see https://bugzilla.mozilla.org/show_bug.cgi?id=1631909

# Accessing BigQuery

There are many methods you can use to access BigQuery: both interactive and programmatic. This document tries to give some basic information and pointers on how to get started with each.

## Table of Contents

<!-- toc -->

## Interfaces

### STMO (`sql.telemetry.mozilla.org`)

All users with access to [STMO](../../tools/stmo.md) can access BigQuery using the following data sources:

- `Telemetry (BigQuery)`
- `Telemetry Search (BigQuery)`

> Note that queries made from STMO are read-only: you cannot create views or tables.

Author: I'm going to leave this one as-is: the "therefore" doesn't quite read right to me as it implies causality rather than identity. Wrt the note part, we can actually use a blockquote which looks a little better than what you suggest, I'll implement this before landing.

### BigQuery Console

> This method requires [BigQuery Access](#BigQuery-Access-Request) to be provisioned.

> The BigQuery console is similar to STMO, but allows write access to views and tables. Some

Author: the "but" is important here

people also prefer its user interface, though note that results that you get from it can
only be shared with others who also have BigQuery access provisioned.

- Visit [GCP BigQuery Console](https://console.cloud.google.com/bigquery)

> - Switch to the project provided to you during your access request e.g `moz-fx-data-bq-<team-name>`

Author: This should actually read "the project that was provided to you in response to your access request"

- Write and run your queries

> Note that if you are trying to query the telemetry dataset, you will need to explicitly specify

Author: I think again this is better done with a blocktext, let me try that locally.

python -c 'from google.cloud import bigquery; print([d.dataset_id for d in bigquery.Client().list_datasets()])'
```

### Spark

#### Locally

> This method requires [BigQuery Access](#BigQuery-Access-Request) to be provisioned.

> If you want to use Spark locally (or via an arbitrary GCP instance in the cloud), we recommend the [Storage API Connector](https://github.com/GoogleCloudPlatform/spark-bigquery-connector) for accessing BigQuery tables in Spark as it is the most modern and actively developed connector. It works well with the BigQuery client library which is useful if you need to run arbitrary SQL queries and load their results into Spark (see the [Databricks](#databricks) section for a link to an example notebook demonstrating this technique).

Author: "you apply the" and "get access to" are not quite right, but agree with the removal of "we". Will make modifications locally.

#### Databricks

[Databricks Notebooks](https://docs.databricks.com/notebooks/index.html) provide an interactive

> computational environment, similar to Jupyter. If you are a Mozilla employee, you should be able to access it via [`sso.mozilla.com/databricks`](https://sso.mozilla.com/databricks).

Reviewer: Is "Jupyter" the correct spelling?

> The `shared_serverless_python3` cluster is configured with shared default GCP credentials that are automatically picked up by BigQuery client libraries. It also has the Spark Storage API Connector library added as seen in the example [Python notebook](https://dbc-caf9527b-e073.cloud.databricks.com/#notebook/141939).

Reviewer: ```suggestion
The `shared_serverless_python3` cluster is configured with shared default GCP credentials that are automatically received by BigQuery client libraries. It also includes the Spark Storage API Connector library that has been added as shown in the example [Python notebook](https://dbc-caf9527b-e073.cloud.databricks.com/#notebook/141939).
```

Author: "that has been added" is not quite right here, actually added should not appear at all (my mistake). Will fix locally

Author: "received" isn't quite right, and also we shouldn't use "added". but 👍 on has->includes; will modify this before landing.

#### Dataproc

> This method requires [BigQuery Access](#BigQuery-Access-Request) to be provisioned.

> Dataproc is Google's managed Spark cluster service. Accessing BigQuery from there will be faster than from Databricks because it does not involve cross-cloud data transfers.

Reviewer: ```suggestion
Dataproc is Google's managed Spark cluster service. Getting access to BigQuery from the Spark cluster service is faster than trying to get access from Databricks because it does not involve cross-cloud data transfers.
```

Author: This isn't correct, in this case we're *accessing* the service, not getting access to it.

python -c 'from google.cloud import bigquery; print([d.dataset_id for d in bigquery.Client().list_datasets()])'
```

### Spark

> This method requires [BigQuery Access](#BigQuery-Access-Request) to be provisioned.

We recommend the [Storage API Connector](https://github.com/GoogleCloudPlatform/spark-bigquery-connector) for accessing
BigQuery tables in Spark as it is the most modern and actively developed connector. It works well with the BigQuery
client library which is useful if you need to run arbitrary SQL queries (see example Databricks notebook) and load their
results into Spark.

> ### Databricks

Reviewer 2: Preserve relative heading level:
```suggestion
#### Databricks
```

Author: This was a deliberate change, as some of the remarks about Spark (e.g. special access provisioning) do not apply to databricks. I'm going to keep this one as-is, if you have objections we can take a look in a follow-up.

Author: Actually, on second look, you were right. I think it was just the section above that needed adjusting. Fixed now (I think)

[Databricks Notebooks](https://docs.databricks.com/notebooks/index.html) provide an interactive

> computational environment, similar to Jupyter. If you are a Mozilla employee, you should be able to access it via [`sso.mozilla.com`](https://sso.mozilla.com).

Reviewer 2: We can put a vanity url here:
```suggestion
computational environment, similar to Jupyter. If you are a Mozilla employee, you should be able to access it via [`sso.mozilla.com/databricks`](https://sso.mozilla.com/databricks).
```

The `shared_serverless_python3` cluster is configured with shared default GCP credentials that are automatically picked up by BigQuery client libraries. It also has the Storage API Connector library added as seen in the example [Python notebook](https://dbc-caf9527b-e073.cloud.databricks.com/#notebook/141939).

> ### Dataproc

Reviewer 2: Maintain relative heading level:
```suggestion
#### Dataproc
```

Author: This makes sense, but it should be moved up (to before) Databricks. I'll move this up in a seperate commit, will have to adjust some prose.

Author: Wound up not doing this, see above.

    --project project-id
```

> The Jupyter URL can be retrieved with the following command:

Reviewer: ```suggestion
You can retrieve the Jupyter URL with the following command:
```

```bash
gcloud beta dataproc clusters describe cluster-name --region=us-west1 --project project-id | grep Jupyter
```

After you've finished your work, it's a good practice to delete your cluster:

```bash
gcloud beta dataproc clusters delete cluster-name --region=us-west1 --project project-id --quiet
```

### Colaboratory

> > This method requires [BigQuery Access](#BigQuery-Access-Request) to be provisioned..

Reviewer: ```suggestion
> This method requires [BigQuery Access](#BigQuery-Access-Request) to be provisioned.
```

[Colaboratory](https://colab.research.google.com) (Colab) is Jupyter notebook environment, managed by Google and running in the cloud. Notebooks are stored in Google Drive and can be shared in a similar way to Google Docs.

> Colab can be used to easily access BigQuery and perform analyses. See the [`Telemetry Hello World` notebook](https://colab.research.google.com/drive/1uXmrPnqzDATiCVH2RNJKD8obIZuofFHx) for an interactive example. Under the hood, it uses the BigQuery API to read and write to BigQuery tables, so access needs to be explicitly provisioned.

Reviewer: ```suggestion
You can use Colab to get access to BigQuery easily and perform analyses. See the [`Telemetry Hello World` notebook](https://colab.research.google.com/drive/1uXmrPnqzDATiCVH2RNJKD8obIZuofFHx) for an interactive example. It uses the BigQuery API to read and write to BigQuery tables. Therefore, access needs to be provisioned explicitly.
```

Author: I think some meaning got lost here-- same problem with the "therefore" and "access" as above. I think I'm going to leave this one as-is.
